---
title: "R Notebook"
output: html_notebook
---
```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = normalizePath("/tank/data/ernesto/analysis/SCT01-10_scRNA_GitCloned")) 
knitr::opts_chunk$set(echo = TRUE)

Sys.setenv(JAVA_HOME="/home/ernesto/java/jdk-11/bin")
Sys.setenv(PATH=paste(Sys.getenv("JAVA_HOME"), Sys.getenv("PATH"), sep=.Platform$path.sep))

#devtools::install_github("broadinstitute/infercnv")
library(infercnv)
library(rjags)

#install.packages("MASS")
library(MASS)

#install.packages("tibble")
library(tibble)

#install.packages("parallelDist")
library(parallelDist)

#install.packages("parallel", force = TRUE, lib = "/Users/genomelab/Library/R/x86_64/4.3/library")
library(parallel)
library(scCustomize)
#include to visualize the plots
# install.packages("devtools")
# devtools::install_github("bmbroom/tsvio", force=TRUE)
# devtools::install_github("bmbroom/NGCHMR", ref="stable", force=TRUE)
# devtools::install_github("broadinstitute/inferCNV_NGCHM", force=TRUE)

#install.packages("ggplot2")
library(ggplot2)

#install.packages("Seurat")
library(Seurat)
library(Matrix)

#install.packages("data.table")
library("data.table")

#install.packages("magrittr") 
#install.packages("dplyr", dependencies = TRUE)    
library(magrittr) 
library(dplyr) 
library(RColorBrewer)

library(biomaRt)
library(ggplot2)
library(data.table)

# Immune (Blue) shades
Immune_shades <- c("#85C1E9", "#5DADE2", "#3498DB", "#2980B9", "#1F618D")

# Neuroectoderm (Red) shades
Neuroectoderm_shades <- c("#E57373", "#D64A4A", "#B53A3A", "#F28C8C", "#A83232", "#F5A09E")

# Epithelia (Green) shades
Epithelia_shades <- c("#A8E6A3", "#7CC47F", "#61B14B", "#88C992", "#4E9835", "#009A44")

# Stroma (Purple) shades
Stroma_shades <- c("#D1A1DA", "#C882D5", "#A259C0", "#8E38A7", "#754C96", "#A67BB4", "#B46BB2", "#6B3583")

# Endothelia (Orange) shades
Endothelia_shades <- c("#F9A14B", "#F47D2D", "#D36E20", "#FFB14C", "#D37D00")

# Combine all shades into one palette
pal1 <- c(
  Epithelia_shades,
  Stroma_shades,
  Endothelia_shades,
  Immune_shades,
  Neuroectoderm_shades
)

# Assign names to the colors
names(pal1) <- c(
  "BPIFB1+_epi",
  "OCT4+_epi",
  "LGR5+_epi",
  "Epithelia_spare",
  "cycl_epi",
  "cilia_epi",  # Extra in case you need it

  "fibroblast",
  "COL12A1+_myofibro",
  "sm_musc",
  "chondrocyte",
  "cycl_mesench",
  "PAX7+_musc_MSC",
  "sk_musc",
  "Stroma_spare2",  # Extra in case you need it

  "vasc_endo",
  "vein_endo",
  "lymph_endo",
  "cycl_endo",
  "Endothelia_spare",  # Extra in case you need it

  "mac",
  "infl_mac",
  "t_cell",
  "mast",
  "nkt",
  "Immune_spare",  # Extra in case you need it

  "GFAP+_astro_radGlia",
  "neuro",
  "oligo",
  "enterochromaffin",
  "CUX2+_cilia_astro_radGlia"
)

# Check the final palette
pal1


pal_broad <- c("Stroma" =  "#9467bd", "Epithelia" = "#2ca02c", "Endothelia" = "#ff7f0e", "Neuroectoderm" = "#d62728", "Immune" = "#1f77b4")
set.seed(12345)

pal_solidcystic <- c("solid" = "#696969", "cystic" = "#D3D3D3")
pal_epithrichpoor <- c("EpitheliaRich" = "orange", "EpitheliaPoor" = "blue", "epithelia_rich" = "orange", "epithelia_poor"= "blue")
```

```{r}
# Analyze NBL scRNA-seq
sobj <- readRDS("data/nucSamples_Harmony_LowResAdded.rds")
metadata <- readRDS("data/MetadataFor_HiResNamedOrdered_SCT01renamed_chromaffinPosChanged_removedRedund_20241120.rds")
sobj@meta.data = metadata
rm(metadata)
sobj$EpitheliaRichPoor = factor(sobj$EpitheliaRichPoor, levels = c("EpitheliaPoor", "EpitheliaRich"))

Idents(sobj) <- sobj$HiResNamedOrdered

Idents(sobj) <- "orig.ident"
sobj <- RenameIdents(sobj, "SCT01" = "epitheliaPoor",
                      "SCT03" = "epitheliaPoor", 
                      "SCT09" = "epitheliaPoor",
                       "SCT10" = "epitheliaRich_OCT4hi", 
                       "SCT02" = "epitheliaRich_OCT4lo",
                     "SCT06" = "epitheliaRich_OCT4lo")
sobj$EpitheliaSplit <- Idents(sobj)
sobj$EpitheliaSplit = factor(sobj$EpitheliaSplit, levels = c("epitheliaPoor", "epitheliaRich_OCT4lo", "epitheliaRich_OCT4hi"))

Idents(sobj) <- sobj$HiResNamedOrdered

sobj
DimPlot_scCustom(sobj, reduction = "umap.harmony", group.by = "HiResNamedOrdered", colors_use = pal1)
```


```{r}
infercnv_sobj_immuneNorm_Xonly_default = readRDS(file ="data/20240824_InferCNV_HiResNamed_xonly/infercnv_run_xonly_sobj_default.RDS")
infercnv_sobj_immuneNorm_default = readRDS(file ="data/20240824_InferCNV_HiResNamed/infercnv_run_sobj_default.RDS")

```





```{r}
infercnv_subclusters.observation_groupings <- read.csv("/tank/data/ernesto/analysis/SCT01-10_scRNA_GitCloned/data/20240824_InferCNV_HiResNamed/outdir_InferCNV/infercnv_subclusters.observation_groupings.txt", sep="")

```


```{r}
sobj = infercnv::add_to_seurat(infercnv_output_path="data/20240824_InferCNV_HiResNamed/outdir_InferCNV",
                                     seurat_obj=sobj, # optional
                                     top_n=20
                                     )


chrXmeta = grep(pattern = "chrX", x = colnames(sobj@meta.data), value = T)
chrXmetaTF = grep(pattern = "has", x = chrXmeta, value = T)
chrXmetaVal = setdiff(chrXmeta, chrXmetaTF)



DimPlot_scCustom(sobj, group.by = c("has_cnv_chrX",  "has_loss_chrX" , "has_dupli_chrX" ), reduction = "umap.harmony", colors_use = c('grey', 'purple'))

pdf("figures/20250110_XChromosome_InferCNV_WITHxaRatios/FeaturePlotsWith_XchrMetadata.pdf", width = 18, height = 12)
DimPlot_scCustom(sobj, group.by = chrXmetaTF, reduction = "umap.harmony", colors_use = c('grey', 'purple'))
FeaturePlot_scCustom(sobj, features = chrXmetaVal, reduction = "umap.harmony")
dev.off()

pdf("figures/20250110_XChromosome_InferCNV_WITHxaRatios/FeaturePlotsWith_XchrMetadata_split.pdf", width = 18, height = 20)
DimPlot_scCustom(sobj, group.by = chrXmetaTF, reduction = "umap.harmony", colors_use = c('grey', 'purple'), split.by = "SolidCystic", split_seurat = T)
FeaturePlot_scCustom(sobj, features = chrXmetaVal, reduction = "umap.harmony", split.by = "SolidCystic")
DimPlot_scCustom(sobj, group.by = chrXmetaTF, reduction = "umap.harmony", colors_use = c('grey', 'purple'), split.by = "EpitheliaRichPoor", split_seurat = T)
FeaturePlot_scCustom(sobj, features = chrXmetaVal, reduction = "umap.harmony", split.by = "EpitheliaRichPoor")
DimPlot_scCustom(sobj, group.by = chrXmetaTF, reduction = "umap.harmony", colors_use = c('grey', 'purple'), split.by = "EpitheliaSplit", split_seurat = T)
FeaturePlot_scCustom(sobj, features = chrXmetaVal, reduction = "umap.harmony", split.by = "EpitheliaSplit")
dev.off()

```

```{r}
DotPlot_scCustom(sobj, features = "proportion_loss_chrY", x_lab_rotate = T, group.by = "orig.ident")
DotPlot_scCustom(sobj, features = "proportion_dupli_chrY", x_lab_rotate = T, group.by = "orig.ident")

DotPlot_scCustom(sobj, features = "proportion_loss_chrX", x_lab_rotate = T, group.by = "orig.ident")
DotPlot_scCustom(sobj, features = "proportion_dupli_chrX", x_lab_rotate = T, group.by = "orig.ident")

DotPlot_scCustom(sobj, features = "has_loss_chrX", x_lab_rotate = T, group.by = "orig.ident")
DotPlot_scCustom(sobj, features = "has_dupli_chrX", x_lab_rotate = T, group.by = "orig.ident")

DotPlot_scCustom(sobj, features = "has_loss_chrY", x_lab_rotate = T, group.by = "orig.ident")
DotPlot_scCustom(sobj, features = "has_dupli_chrY", x_lab_rotate = T, group.by = "orig.ident")


dotplotfeature = c( "proportion_loss_chrX","proportion_dupli_chrX", "has_loss_chrX", "has_dupli_chrX", "proportion_loss_chrY",  "proportion_dupli_chrY", "has_loss_chrY", "has_dupli_chrY")

DotPlot_scCustom(sobj, features = dotplotfeature, x_lab_rotate = T, group.by = "orig.ident")
DotPlot_scCustom(sobj, features = c("has_loss_chrX", "has_dupli_chrX", "proportion_loss_chrY",  "proportion_dupli_chrY", "has_loss_chrY", "has_dupli_chrY"), x_lab_rotate = T, group.by = "orig.ident")


```
```{r}
VlnPlot_scCustom(sobj, features = "XIST", pt.size = 0, group.by = "orig.ident", split.by = "BroadCategory", colors_use = pal_broad)
```




```{r}
# Connect to Ensembl (GRCh38/hg38)
ensembl <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
# Retrieve gene coordinates
gene_coord <- unique(as.data.frame(getBM(attributes = c("hgnc_symbol", "chromosome_name", "start_position", "end_position","strand"), mart = ensembl)))

# Remove entries with empty gene names
gene_coord <- gene_coord[gene_coord$hgnc_symbol != "",]


# Extract the expression matrix
mtx <- sobj@assays$RNA$data

# Extract cell cluster or cell type info
cell_clusters <- data.frame(index = rownames(sobj@meta.data), 
                            louvain = sobj@meta.data$HiResNamedOrdered)


# Get X chromosome genes
x_genes <- gene_coord$hgnc_symbol[gene_coord$chromosome_name == "X"]
x_genes <- x_genes[x_genes %in% rownames(mtx)]  # Only keep genes present in the expression data

# Get Y chromosome genes
y_genes <- gene_coord$hgnc_symbol[gene_coord$chromosome_name == "Y"]
y_genes <- y_genes[y_genes %in% rownames(mtx)]  # Only keep genes present in the expression data


# Get autosomal genes (chromosomes 1-22)
autosomal_genes <- gene_coord$hgnc_symbol[gene_coord$chromosome_name %in% as.character(1:22)]
autosomal_genes <- autosomal_genes[autosomal_genes %in% rownames(mtx)]  # Only keep genes present in the expression data



# Sum expression for X-chromosome genes
x_sum <- colSums(mtx[x_genes, ])

# Sum expression for autosomal genes
autosomal_sum <- colSums(mtx[autosomal_genes, ])

# Calculate the ratio of X-chromosome to autosomal gene expression for each cell
x_to_autosome_ratio <- x_sum / autosomal_sum

# Merge the ratio with cell cluster data
cell_clusters$X_to_Autosome_Ratio <- x_to_autosome_ratio
```

```{r}
# Extract XIST and XACT expression values from the Seurat object
XIST_expression <- sobj@assays$RNA$data["XIST", ]
XACT_expression <- sobj@assays$RNA$data["XACT", ]

# Define threshold for "expressing" - you can set the threshold based on your data
# Here, we'll assume that an expression value > 0 means the gene is expressed.
XIST_expressed <- XIST_expression > 0
XACT_expressed <- XACT_expression > 0

```

```{r}
# Create buckets for XIST
sobj$XIST_expr_metadata <- ifelse(XIST_expressed, "XIST_Expressed", "XIST_Not_Expressed")

# Create buckets for XACT
sobj$XACT_expr_metadata <- ifelse(XACT_expressed, "XACT_Expressed", "XACT_Not_Expressed")

```

```{r}
# Assuming x_to_autosome_ratio has already been computed and added to the Seurat object as described previously:
sobj$X_to_Autosome_Ratio <- x_to_autosome_ratio

# Subset the data by XIST expression
XIST_group <- sobj@meta.data %>% dplyr::select(XIST_expr_metadata, X_to_Autosome_Ratio)

# Subset the data by XACT expression
XACT_group <- sobj@meta.data %>% dplyr::select(XACT_expr_metadata, X_to_Autosome_Ratio)
```



```{r}
# Extract the metadata to get orig.ident (to separate male/female)
metadata <- sobj@meta.data

# Identify male and female samples
male_cells <- metadata$orig.ident == "SCT06"
female_cells <- metadata$orig.ident != "SCT06"  # Everything else assumed to be female

# Define groups for XIST expression
sobj$XIST_groups <- ifelse(male_cells, "Male",
                          ifelse(XIST_expressed & female_cells, "Female_XIST_Expressed", "Female_XIST_Not_Expressed"))

# Define groups for XACT expression
sobj$XACT_groups <- ifelse(male_cells, "Male",
                          ifelse(XACT_expressed & female_cells, "Female_XACT_Expressed", "Female_XACT_Not_Expressed"))

# Check the new groupings
table(sobj$XIST_groups)
table(sobj$XACT_groups)

```

```{r}
# Subset the data for XIST groups and X-to-autosome ratio
XIST_groups_data <- sobj@meta.data %>% dplyr::select(XIST_groups, X_to_Autosome_Ratio)

# Subset the data for XACT groups and X-to-autosome ratio
XACT_groups_data <- sobj@meta.data %>% dplyr::select(XACT_groups, X_to_Autosome_Ratio)
```


```{r}
# Plot X-to-Autosomal ratio for XIST groups
p1 = ggplot(XIST_groups_data, aes(x = XIST_groups, y = X_to_Autosome_Ratio, fill = XIST_groups)) +
  geom_violin() +
  ylab("X-to-Autosomal Ratio") +
  xlab("XIST Groups") +
  theme_minimal() + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + geom_hline(yintercept = 0.036)

# Plot X-to-Autosomal ratio for XACT groups
p2 = ggplot(XACT_groups_data, aes(x = XACT_groups, y = X_to_Autosome_Ratio, fill = XACT_groups)) +
  geom_violin() +
  ylab("X-to-Autosomal Ratio") +
  xlab("XACT Groups") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + geom_hline(yintercept = 0.036)


p1 | p2

```


```{r}
FeaturePlot_scCustom(sobj, features = c("X_to_Autosome_Ratio"), reduction = "umap.harmony", na_cutoff = 0.036)
FeaturePlot_scCustom(sobj, features = c("X_to_Autosome_Ratio"), reduction = "umap.harmony", split.by = "SolidCystic")
FeaturePlot_scCustom(sobj, features = c("X_to_Autosome_Ratio"), reduction = "umap.harmony", split.by = "EpitheliaRichPoor")
FeaturePlot_scCustom(sobj, features = c("X_to_Autosome_Ratio"), reduction = "umap.harmony", split.by = "orig.ident", num_columns = 3)
```
```{r}
sobj_female <- subset(sobj, subset = orig.ident != "SCT06")

```


```{r}
FeatureScatter_scCustom(seurat_object = sobj, feature1 = "X_to_Autosome_Ratio", feature2 = "XIST", colors_use = pal_broad, group.by = "BroadCategory", pt.size = 1, split.by = "HiResNamed", num_columns = 5, split_seurat = F, plot.cor = T) & geom_vline(xintercept = 0.036) & geom_hline(yintercept = 1)

FeatureScatter_scCustom(seurat_object = sobj, feature1 = "X_to_Autosome_Ratio", feature2 = "XACT", colors_use = pal_broad, group.by = "BroadCategory", pt.size = 1, split.by = "HiResNamed", num_columns = 5, split_seurat = F, plot.cor = T) & geom_vline(xintercept = 0.036) & geom_hline(yintercept = 1)

FeatureScatter_scCustom(seurat_object = sobj, feature1 = "X_to_Autosome_Ratio", feature2 = "XIST", colors_use = pal_broad, group.by = "BroadCategory", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T) & geom_vline(xintercept = 0.036) & geom_hline(yintercept = 1)
```
```{r}
FeatureScatter_scCustom(seurat_object = sobj_female, feature1 = "X_to_Autosome_Ratio", feature2 = "XIST", colors_use = pal_broad, group.by = "BroadCategory", pt.size = 1, split.by = "HiResNamed", num_columns = 5, split_seurat = F, plot.cor = T) & geom_vline(xintercept = 0.036) & geom_hline(yintercept = 1)

FeatureScatter_scCustom(seurat_object = sobj_female, feature1 = "X_to_Autosome_Ratio", feature2 = "XACT", colors_use = pal_broad, group.by = "BroadCategory", pt.size = 1, split.by = "HiResNamed", num_columns = 5, split_seurat = F, plot.cor = T) & geom_vline(xintercept = 0.036) & geom_hline(yintercept = 1)

FeatureScatter_scCustom(seurat_object = sobj_female, feature1 = "X_to_Autosome_Ratio", feature2 = "XIST", colors_use = pal_broad, group.by = "BroadCategory", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T) & geom_vline(xintercept = 0.036) & geom_hline(yintercept = 1)
```

```{r}
VlnPlot_scCustom(sobj, features = "X_to_Autosome_Ratio", group.by = "HiResNamedOrdered", split.by = "orig.ident", pt.size = 0) + geom_hline(yintercept = c(0.05, 0.025),linetype='dotted') + geom_hline(yintercept = 0.036)
```

```{r}
FeatureScatter_scCustom(seurat_object = sobj, feature1 = "X_to_Autosome_Ratio", feature2 = "proportion_dupli_chrX", colors_use = pal_broad, group.by = "BroadCategory", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T) & geom_vline(xintercept = 0.036) & geom_hline(yintercept = .5)
FeatureScatter_scCustom(seurat_object = sobj, feature1 = "X_to_Autosome_Ratio", feature2 = "proportion_dupli_chrX", colors_use = pal1, group.by = "HiResNamedOrdered", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T) & geom_vline(xintercept = 0.036) & geom_hline(yintercept = .5)
FeatureScatter_scCustom(seurat_object = sobj, feature1 = "X_to_Autosome_Ratio", feature2 = "proportion_dupli_chrX", group.by = "orig.ident", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T) & geom_vline(xintercept = 0.036) & geom_hline(yintercept = .5)
```


```{r}
FeatureScatter_scCustom(seurat_object = sobj_female, feature1 = "X_to_Autosome_Ratio", feature2 = "proportion_dupli_chrX", colors_use = pal_broad, group.by = "BroadCategory", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T) & geom_vline(xintercept = 0.036) & geom_hline(yintercept = .5)
FeatureScatter_scCustom(seurat_object = sobj_female, feature1 = "X_to_Autosome_Ratio", feature2 = "proportion_dupli_chrX", colors_use = pal1, group.by = "HiResNamedOrdered", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T) & geom_vline(xintercept = 0.036) & geom_hline(yintercept = .5)
FeatureScatter_scCustom(seurat_object = sobj_female, feature1 = "X_to_Autosome_Ratio", feature2 = "proportion_dupli_chrX", group.by = "orig.ident", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T) & geom_vline(xintercept = 0.036) & geom_hline(yintercept = .5)
```


```{r}
library(dplyr)

# Add a new metadata column based on the conditions
sobj@meta.data <- sobj@meta.data %>%
  mutate(Xa_called = case_when(
    X_to_Autosome_Ratio > 0.036 & proportion_dupli_chrX > 0.5 ~ "XaXa",
    X_to_Autosome_Ratio < 0.036 & proportion_dupli_chrX < 0.5 ~ "XaXi",
    TRUE ~ "Unknown"
  ))

# Check the distribution of new metadata
table(sobj@meta.data$Xa_called)


# Add a new metadata column based on the conditions
sobj_female@meta.data <- sobj_female@meta.data %>%
  mutate(Xa_called = case_when(
    X_to_Autosome_Ratio > 0.036 & proportion_dupli_chrX > 0.5 ~ "XaXa",
    X_to_Autosome_Ratio < 0.036 & proportion_dupli_chrX < 0.5 ~ "XaXi",
    TRUE ~ "Unknown"
  ))

# Check the distribution of new metadata
table(sobj_female@meta.data$Xa_called)



```




```{r}
FeatureScatter_scCustom(seurat_object = sobj, feature1 = "X_to_Autosome_Ratio", feature2 = "has_dupli_chrX", group.by = "orig.ident", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T) & geom_vline(xintercept = 0.036) & geom_hline(yintercept = .5)
FeatureScatter_scCustom(seurat_object = sobj, feature1 = "XIST", feature2 = "proportion_dupli_chrX", group.by = "orig.ident", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T) & geom_vline(xintercept = 0.036) & geom_hline(yintercept = .5)
```

```{r}
sobj$orig.ident = factor(sobj$orig.ident, levels = c('SCT01','SCT03','SCT10','SCT06','SCT02','SCT09'))
FeaturePlot_scCustom(sobj, features = "XIST", reduction = "umap.harmony", split.by = "orig.ident", num_columns = 1)
FeaturePlot_scCustom(sobj, features = "XACT", reduction = "umap.harmony", split.by = "orig.ident", num_columns = 1)
```


```{r}
# Load necessary libraries
library(scCustomize)
library(patchwork)

# Define the features to plot
features_to_plot <- c("proportion_loss_chrX", "proportion_dupli_chrX")

# Generate individual DotPlots separately
p1 <- DotPlot_scCustom(
    subset(sobj, orig.ident == "SCT10"), 
    features = features_to_plot, 
    group.by = "HiResNamedOrdered", 
    x_lab_rotate = TRUE
) + ggtitle("SCT10") + theme(legend.position = "none")

p2 <- DotPlot_scCustom(
    subset(sobj, orig.ident == "SCT01"), 
    features = features_to_plot, 
    group.by = "HiResNamedOrdered", 
    x_lab_rotate = TRUE
) + ggtitle("SCT01") + theme(legend.position = "none")

p3 <- DotPlot_scCustom(
    subset(sobj, orig.ident == "SCT03"), 
    features = features_to_plot, 
    group.by = "HiResNamedOrdered", 
    x_lab_rotate = TRUE
) + ggtitle("SCT03") + theme(legend.position = "none")

p4 <- DotPlot_scCustom(
    subset(sobj, orig.ident == "SCT09"), 
    features = features_to_plot, 
    group.by = "HiResNamedOrdered", 
    x_lab_rotate = TRUE
) + ggtitle("SCT09") + theme(legend.position = "none")

p5 <- DotPlot_scCustom(
    subset(sobj, orig.ident == "SCT02"), 
    features = features_to_plot, 
    group.by = "HiResNamedOrdered", 
    x_lab_rotate = TRUE
) + ggtitle("SCT02") + theme(legend.position = "none")

p6 <- DotPlot_scCustom(
    subset(sobj, orig.ident == "SCT06"), 
    features = features_to_plot, 
    group.by = "HiResNamedOrdered", 
    x_lab_rotate = TRUE
) + ggtitle("SCT06") + theme(legend.position = "none")

# Arrange all plots in a single row using patchwork
 (p1 + p2 + p3 + p4 + p5) + plot_layout(ncol = 5)

```

```{r}
# Load necessary libraries
library(scCustomize)
library(patchwork)

# Define the features to plot
features_to_plot <- c("X_to_Autosome_Ratio")

# Generate individual DotPlots separately
p1 <- DotPlot_scCustom(
    subset(sobj, orig.ident == "SCT10"), 
    features = features_to_plot, 
    group.by = "HiResNamedOrdered", 
    x_lab_rotate = TRUE
) + ggtitle("SCT10") + theme(legend.position = "none")

p2 <- DotPlot_scCustom(
    subset(sobj, orig.ident == "SCT01"), 
    features = features_to_plot, 
    group.by = "HiResNamedOrdered", 
    x_lab_rotate = TRUE
) + ggtitle("SCT01") + theme(legend.position = "none")

p3 <- DotPlot_scCustom(
    subset(sobj, orig.ident == "SCT03"), 
    features = features_to_plot, 
    group.by = "HiResNamedOrdered", 
    x_lab_rotate = TRUE
) + ggtitle("SCT03") + theme(legend.position = "none")

p4 <- DotPlot_scCustom(
    subset(sobj, orig.ident == "SCT09"), 
    features = features_to_plot, 
    group.by = "HiResNamedOrdered", 
    x_lab_rotate = TRUE
) + ggtitle("SCT09") + theme(legend.position = "none")

p5 <- DotPlot_scCustom(
    subset(sobj, orig.ident == "SCT02"), 
    features = features_to_plot, 
    group.by = "HiResNamedOrdered", 
    x_lab_rotate = TRUE
) + ggtitle("SCT02") + theme(legend.position = "none")

p6 <- DotPlot_scCustom(
    subset(sobj, orig.ident == "SCT06"), 
    features = features_to_plot, 
    group.by = "HiResNamedOrdered", 
    x_lab_rotate = TRUE
) + ggtitle("SCT06") + theme(legend.position = "none")

# Arrange all plots in a single row using patchwork
 (p1 + p2 + p3 + p4 + p5) + plot_layout(ncol = 5)

```


```{r}
# Load necessary libraries
library(pheatmap)
library(dplyr)
library(RColorBrewer)

# Extract relevant data from Seurat object
heatmap_data <- sobj_female@meta.data %>%
  dplyr::select(HiResNamedOrdered, orig.ident, X_to_Autosome_Ratio, BroadCategory) %>%
  group_by(HiResNamedOrdered, orig.ident) %>%
  summarise(X_to_Autosome_Ratio = mean(X_to_Autosome_Ratio, na.rm = TRUE)) %>%
  pivot_wider(names_from = orig.ident, values_from = X_to_Autosome_Ratio, values_fill = NA) %>%
  as.data.frame()

# Convert HiResNamedOrdered to row names
rownames(heatmap_data) <- heatmap_data$HiResNamedOrdered
heatmap_data$HiResNamedOrdered <- NULL  # Remove the column

# Fix BroadCategory mapping: Ensure it includes only matching cell types
broad_category_mapping <- sobj_female@meta.data %>%
  dplyr::select(HiResNamedOrdered, BroadCategory) %>%
  distinct() %>%
  filter(HiResNamedOrdered %in% rownames(heatmap_data)) %>%  # Keep only matching rows
  as.data.frame()

# Ensure row order in broad_category_mapping matches heatmap_data
broad_category_mapping <- broad_category_mapping[match(rownames(heatmap_data), broad_category_mapping$HiResNamedOrdered), ]
broad_category_mapping$HiResNamedOrdered <- NULL  # Remove redundant column

# Create annotation for BroadCategory
annotation_row <- data.frame(BroadCategory = broad_category_mapping$BroadCategory)
rownames(annotation_row) <- rownames(heatmap_data)

# Generate heatmap with pheatmap
pheatmap(
  as.matrix(heatmap_data),
  cluster_rows = FALSE,  # Keep order based on HiResNamedOrdered
  cluster_cols = TRUE,   # Cluster samples
  annotation_row = annotation_row,  # Add BroadCategory colors
  annotation_colors = list(BroadCategory = pal_broad),
  show_rownames = TRUE,
  show_colnames = TRUE,
  fontsize_row = 10,
  fontsize_col = 10
)


```


```{r}
saveRDS(object = sobj@meta.data,"data/20250110_XChromosome_InferCNV_WITHxaRatios/metadataForXchr.rds")

```




# Trying out correlation of the X-
```{r eval=FALSE, include=FALSE}
# Load required libraries
library(Seurat)
library(dplyr)
library(stats)
library(qvalue)
# Extract expression matrix (log-normalized counts)
expression_matrix <- GetAssayData(sobj, slot = "data")
# Extract X score
x_score <- sobj$proportion_dupli_chrX
# Function to compute correlation and p-values for each gene
compute_gene_correlations <- function(expr_matrix, score_vector) {
  cor_results <- apply(expr_matrix, 1, function(gene_expr) {
    cor.test(gene_expr, score_vector, method = "kendall", use = "complete.obs") #KENDALL RUNS TOO LONG!!!
  })
  # Extract correlation values and p-values
  cor_values <- sapply(cor_results, function(x) x$estimate)
  p_values <- sapply(cor_results, function(x) x$p.value)
  # Adjust p-values for multiple testing using Benjamini-Hochberg FDR correction
  adj_p_values <- p.adjust(p_values, method = "BH")
  # Convert to data frame and rank by absolute correlation
  cor_df <- data.frame(
    gene = rownames(expr_matrix),
    correlation = cor_values,
    p_value = p_values,
    adj_p_value = adj_p_values
  ) %>%
    arrange(desc(abs(correlation)))
  return(cor_df)
}
# Run correlation analysis
gene_correlations <- compute_gene_correlations(expression_matrix, x_score)
# Show top 20 genes with strongest correlation
head(gene_correlations, 20)
# Comparison: Pearson vs. Spearman correlation
# Switched to Spearman, which is rank-based and better for detecting monotonic trends.
# Spearman is more robust to outliers and captures non-linear relationships.
# Pearson should still be considered if linearity is assumed.


```


```{r}
# Load required libraries
library(Seurat)
library(dplyr)
library(stats)
library(qvalue)


# Extract expression matrix (log-normalized counts)
expression_matrix <- GetAssayData(sobj_female, layer = "data")
# Extract X score
x_score <- sobj_female$proportion_dupli_chrX
# Function to compute correlation and p-values for each gene
compute_gene_correlations <- function(expr_matrix, score_vector) {
  cor_results <- apply(expr_matrix, 1, function(gene_expr) {
    cor.test(gene_expr, score_vector, method = "spearman", exact = FALSE, use = "complete.obs")
  })
  
  # Extract correlation values and p-values
  cor_values <- sapply(cor_results, function(x) x$estimate)
  p_values <- sapply(cor_results, function(x) x$p.value)
  
  # Adjust p-values for multiple testing using Benjamini-Hochberg FDR correction
  adj_p_values <- p.adjust(p_values, method = "BH")
  
  # Compute q-values using qvalue package
  q_values <- qvalue(p_values)$qvalues
  
  # Convert to data frame and rank by absolute correlation
  cor_df <- data.frame(
    gene = rownames(expr_matrix),
    correlation = cor_values,
    p_value = p_values,
    adj_p_value = adj_p_values,
    q_value = q_values
  ) %>%
    arrange(desc(abs(correlation)))
  
  return(cor_df)
}
# Run correlation analysis
gene_correlations_dupliX <- compute_gene_correlations(expression_matrix, x_score)
# Show top 20 genes with strongest correlation
head(gene_correlations_dupliX, 20)
hist(gene_correlations_dupliX$p_value)
hist(gene_correlations_dupliX$adj_p_value)
hist(gene_correlations_dupliX$q_value)

```


```{r}
hist(x_score, breaks = 100, main = "Histogram of X Score", xlab = "X Score", col = "lightblue")

```

Let's take a look at the correlating genes. Then remove the X-Chromosome genes

```{r}
sign_gene_correlations = gene_correlations %>% 
  dplyr::filter(adj_p_value < 0.05 & q_value < 0.05) %>% 
  dplyr::filter(!gene %in% x_genes) %>%
  dplyr::arrange(desc(correlation))

```




```{r}
# Load required libraries
library(Seurat)
library(dplyr)
library(stats)
library(qvalue)


# Extract expression matrix (log-normalized counts)
expression_matrix <- GetAssayData(sobj_female, layer = "data")
# Extract X score
x_score <- sobj_female$X_to_Autosome_Ratio
# Function to compute correlation and p-values for each gene
compute_gene_correlations <- function(expr_matrix, score_vector) {
  cor_results <- apply(expr_matrix, 1, function(gene_expr) {
    cor.test(gene_expr, score_vector, method = "spearman", exact = F, use = "complete.obs")
  })
  
  # Extract correlation values and p-values
  cor_values <- sapply(cor_results, function(x) x$estimate)
  p_values <- sapply(cor_results, function(x) x$p.value)
  
  # Adjust p-values for multiple testing using Benjamini-Hochberg FDR correction
  adj_p_values <- p.adjust(p_values, method = "BH")
  
  # Compute q-values using qvalue package
  q_values <- qvalue(p_values)$qvalues
  
  # Convert to data frame and rank by absolute correlation
  cor_df <- data.frame(
    gene = rownames(expr_matrix),
    correlation = cor_values,
    p_value = p_values,
    adj_p_value = adj_p_values,
    q_value = q_values
  ) %>%
    arrange(desc(abs(correlation)))
  
  return(cor_df)
}
# Run correlation analysis
gene_correlations <- compute_gene_correlations(expression_matrix, x_score)
# Show top 20 genes with strongest correlation
head(gene_correlations, 20)
hist(gene_correlations$p_value)
hist(gene_correlations$adj_p_value)
hist(gene_correlations$q_value)
```





```{r}
# Prepare a ranked gene list for GSEA
gene_list <- gene_correlations$correlation
names(gene_list) <- gene_correlations$gene  # Assign gene names
gene_list <- sort(gene_list, decreasing = TRUE)  # Sort genes by log2FoldChange
```


```{r}
library(clusterProfiler)
library(org.Hs.eg.db)
# Run GSEA for GO terms
gsea_go <- gseGO(
  geneList = gene_list,
  OrgDb = org.Hs.eg.db,
  keyType = "SYMBOL",
  ont = "ALL",  # Use "ALL" for biological process, cellular component, and molecular function
  minGSSize = 10,
  maxGSSize = 500,
  pvalueCutoff = 0.05,
  verbose = TRUE,
  #scoreType = "pos",  # Adjust if needed
  nPermSimple = 10000  # Increase if needed
)

```


What genes are in the x_genes AND in our correlation???
```{r}

significant_genes = gene_correlations %>%
  filter(adj_p_value < 0.05, q_value < 0.05)  # Assuming q-value is same as adj_p_value

print("X-genes that significantly correlate negatively to X/A ratio")
intersect(x_genes, significant_genes[significant_genes$correlation < 0, "gene"])
print("X-genes that significantly correlate positively to X/A ratio")
intersect(x_genes, significant_genes[significant_genes$correlation > 0, "gene"])


```


```{r}
library(enrichplot)
View(gsea_go@result)
# Filter GSEA results for downregulated terms
x_dn <- filter(gsea_go, p.adjust < 0.05, NES < 0, qvalue < 0.05, enrichmentScore < -0.4)

# Calculate pairwise term similarity for downregulated terms
x_dn_sim <- pairwise_termsim(x_dn, showCategory = nrow(x_dn))
```


```{r}
# Create an emapplot for the top 40 downregulated terms
plot_dn_40 <- emapplot(x_dn_sim, showCategory = 40) + 
    ggtitle("Negative Correlation to X/A ratio - Downregulated NES < 0") +
    theme(plot.title = element_text(hjust = 0.5, size = 16, face = "bold"))
print(plot_dn_40)

# Simplify the results based on NES and create another emapplot
x_dn_simple <- simplify(x_dn_sim, cutoff = 0.7, by = "NES", select_fun = min)
plot_dn_simplified <- emapplot(x_dn_simple) + 
    ggtitle("Simplified GSEA - Negative Correlation to X/A ratio - Downregulated NES < 0 - Lowest NES") +
    theme(plot.title = element_text(hjust = 0.5, size = 16, face = "bold"))
print(plot_dn_simplified)

x_dn_simple_2 <- simplify(x_dn_sim, cutoff = 0.7, by="p.adjust", select_fun=min)
plot_dn_simplified2 <- emapplot(x_dn_simple) + 
    ggtitle("Simplified GSEA - Negative Correlation to X/A ratio - Downregulated NES < 0 - pAdj") +
    theme(plot.title = element_text(hjust = 0.5, size = 16, face = "bold"))
print(plot_dn_simplified2)


treeplot(x_dn_simple_2, offset.params = list(bar_tree = rel(5)), showCategory = 20)
```



```{r}
# Filter GSEA results for downregulated terms
x_up <- filter(gsea_go, p.adjust < 0.05, NES > 0, qvalue < 0.05, enrichmentScore > 0.4)

# Calculate pairwise term similarity for downregulated terms
x_up_sim <- pairwise_termsim(x_up, showCategory = nrow(x_up))
```


```{r}
# Create an emapplot for the top 40 downregulated terms
plot_up_40 <- emapplot(x_up_sim, showCategory = 40) + 
    ggtitle("Positive Correlation to X/A ratio - Upregulated NES > 0") +
    theme(plot.title = element_text(hjust = 0.5, size = 16, face = "bold"))
print(plot_up_40)

# Simplify the results based on NES and create another emapplot
x_up_simple <- simplify(x_up_sim, cutoff = 0.7, by = "NES", select_fun = max)
plot_up_simplified <- emapplot(x_up_simple) + 
    ggtitle("Simplified GSEA - Positive Correlation to X/A ratio - Upregulated NES > 0 - Lowest NES") +
    theme(plot.title = element_text(hjust = 0.5, size = 16, face = "bold"))
print(plot_up_simplified)

x_up_simple_2 <- simplify(x_up_sim, cutoff = 0.7, by="p.adjust", select_fun=min)
plot_up_simplified2 <- emapplot(x_up_simple) + 
    ggtitle("Simplified GSEA - Positive Correlation to X/A ratio - Upregulated NES > 0 - pAdj") +
    theme(plot.title = element_text(hjust = 0.5, size = 16, face = "bold"))
print(plot_up_simplified2)



treeplot(x_up_simple_2, offset.params = list(bar_tree = rel(5)), showCategory = 20)

```




```{r}
# Prepare a ranked gene list for GSEA
gene_list_dupliX <- gene_correlations_dupliX$correlation
names(gene_list_dupliX) <- gene_correlations_dupliX$gene  # Assign gene names
gene_list_dupliX <- sort(gene_list_dupliX, decreasing = TRUE)  # Sort genes by log2FoldChange
```


```{r}
library(clusterProfiler)
library(org.Hs.eg.db)
# Run GSEA for GO terms
gsea_go_dupliX <- gseGO(
  geneList = gene_list_dupliX,
  OrgDb = org.Hs.eg.db,
  keyType = "SYMBOL",
  ont = "ALL",  # Use "ALL" for biological process, cellular component, and molecular function
  minGSSize = 10,
  maxGSSize = 500,
  pvalueCutoff = 0.05,
  verbose = TRUE,
  #scoreType = "pos",  # Adjust if needed
  nPermSimple = 10000  # Increase if needed
)
```


Wz
```{r}

```



```{r}
FeatureScatter_scCustom(seurat_object = sobj_female, feature1 = "LDLR", feature2 = "X_to_Autosome_Ratio", colors_use = pal_broad, group.by = "BroadCategory", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T)
FeatureScatter_scCustom(seurat_object = sobj_female, feature1 = "LDLR", feature2 = "proportion_dupli_chrX", colors_use = pal_broad, group.by = "BroadCategory", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T)
FeatureScatter_scCustom(seurat_object = sobj_female, feature1 = "XIST", feature2 = "X_to_Autosome_Ratio", colors_use = pal_broad, group.by = "BroadCategory", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T)
FeatureScatter_scCustom(seurat_object = sobj_female, feature1 = "XIST", feature2 = "proportion_dupli_chrX", colors_use = pal_broad, group.by = "BroadCategory", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T)

FeatureScatter_scCustom(seurat_object = sobj_female, feature1 = "FTX", feature2 = "X_to_Autosome_Ratio", colors_use = pal_broad, group.by = "BroadCategory", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T)
FeatureScatter_scCustom(seurat_object = sobj_female, feature1 = "FTX", feature2 = "proportion_dupli_chrX", colors_use = pal_broad, group.by = "BroadCategory", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T)
FeatureScatter_scCustom(seurat_object = sobj_female, feature1 = "EDA", feature2 = "X_to_Autosome_Ratio", colors_use = pal_broad, group.by = "BroadCategory", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T)
FeatureScatter_scCustom(seurat_object = sobj_female, feature1 = "EDA", feature2 = "proportion_dupli_chrX", colors_use = pal_broad, group.by = "BroadCategory", pt.size = 1, num_columns = 5, split_seurat = F, plot.cor = T)
```


```{r}
# Compute -log10(adjusted p-value)
gene_correlations <- gene_correlations %>%
  mutate(logP = -log10(adj_p_value))

# Define key genes to highlight
key_genes <- c("GPC3", "NKD1", "MLLT3", "TIAM1", "WNT5B", "RSPO3", "SFRP1", 
               "DAAM2", "NPHP3", "ZNRF3", "WNT5A", "ANKRD6", "DACT1")

# Create plot
ggplot(gene_correlations, aes(x = correlation, y = reorder(gene, correlation))) +
  geom_point(aes(size = logP, color = gene %in% key_genes), alpha = 0.7) +
  scale_color_manual(values = c("blue", "red"), labels = c("Other Genes", "Key Genes")) +
  scale_size_continuous(range = c(2, 8)) +
  labs(title = "Significant Gene Correlations with X to Autosome Ratio",
       x = "Correlation with X to Autosome Ratio",
       y = "Genes",
       color = "Gene Type",
       size = "-log10(Adjusted P-Value)") +
  theme_minimal()
```

```{r}
# Define selected genes to label
selected_genes <- c("XIST", "DCC", "TSIX")

# Compute -log10(adjusted p-value) and define colors
gene_correlations <- gene_correlations %>%
  mutate(logP = -log10(adj_p_value),
         sig_color = ifelse(adj_p_value < 0.01, "red", "blue"),
         gene_label = ifelse(gene %in% selected_genes, gene, ""))  # Label only selected genes

# Create plot with color based on adj_p_value
ggplot(gene_correlations, aes(x = correlation, y = reorder(gene, correlation))) +
  geom_point(aes(size = , color = sig_color), alpha = 0.7) +  # Color based on significance
  scale_color_identity() +  # Use raw colors (red, blue) without a legend
  scale_size_continuous(range = c(2, 8)) +
  scale_y_discrete(labels = function(x) ifelse(x %in% selected_genes, x, "")) +  # Show only selected genes
  labs(title = "Significant Gene Correlations with X to Autosome Ratio",
       x = "Correlation with X to Autosome Ratio",
       y = "Genes",
       size = "-log10(Adjusted P-Value)") +
  theme_classic()
########

# Load necessary libraries
library(ggplot2)
library(ggrepel)

# Define selected genes to highlight
selected_genes <- c("XIST", "TSIX", "WNT5B", "FGFR2", "PBX1", "TBX5")  # Add your genes of interest here


# Compute -log10(adjusted p-value) and prepare colors
gene_correlations <- gene_correlations %>%
    mutate(logP = -log10(adj_p_value),  # Convert to log-scale for better visibility
           gene_label = ifelse(gene %in% selected_genes, gene, NA))  # Only label selected genes

# Create plot
ggplot(gene_correlations, aes(x = correlation, y = reorder(gene, correlation))) +
    geom_point(aes(color = adj_p_value), size = 3, alpha = 0.8) +  # Fixed size, transparency for visibility
    scale_color_gradient(low = "red", high = "blue", limits = c(0, 0.01)) +  # Red = most sig, Blue = p=0.05
    geom_text_repel(aes(label = gene_label), 
                    box.padding = 0.5,  # Space around labels
                    point.padding = 0.2,  # Space between dot and label
                    segment.color = "black",  # Color of connecting line
                    segment.size = 0.5,  # Line thickness
                    max.overlaps = Inf) +  # Always show selected genes
    labs(title = "Gene Correlations with X to Autosome Ratio",
         x = "Correlation with X to Autosome Ratio",
         y = NULL,  # Remove y-axis label
         color = "Adjusted P-Value") +  # Legend title for color
    theme_classic() +
    theme(axis.text.y = element_blank(),  # Remove y-axis gene names
          axis.ticks.y = element_blank())  # Remove y-axis ticks

```



