# Single-cell and spatial profiling of human sacrococcygeal teratomas reveals stratifications in cell type composition and X-chromosome inactivation

Ernesto J. Rojas, Krinio Giannikou, ..., Tippi MacKenzie*, Diana Laird*
  
*corresponding author

[biorXiv...no link yet]()


## Abstract
Sacrococcygeal teratomas (SCTs) the most common tumors in newborns, are associated with significant perinatal morbidity. For reasons not yet understood, they occur more frequently in patients with XX chromosomes than in those with XY chromosomes. Because SCTs lack severity biomarkers and exhibit multilineage differentiation, which has complicated their classification and clinical prognostic decisions. We characterized inter-tumor heterogeneity of six postnatal and two prenatal SCT samples by single-nuclei RNA-seq and spatial transcriptomics to delineate the cellular composition of SCTs and identify molecular features associated with their variability. We identified five major lineages: stroma, epithelia, endothelial, neuroectoderm, and immune cells. Although SCTs are thought to originate from pluripotent or primordial germ cells, we did not detect these populations in any tumors. Our findings established a novel stratification based on the presence of epithelial cells (“epithelia-rich” or “epithelia-poor”). Among XX tumors, a subset of cells harbored two active X-chromosomes (XaXa-like). These XaXa-like cells exhibited elevated expression of neuronal and developmental morphogenesis genes, while cells with a single active X showed immune and inflammatory biases. This study provides the first spatial and single-cell atlas of SCTs, revealing distinct cellular subtypes and transcriptional programs. Our findings highlight the potential role of X-chromosome inactivation and immune signaling in SCT biology, offering new avenues for diagnostic and therapeutic development.

## Data
SCT scRNA-seq, snRNA-seq, and spatial transcriptomic data generated for this project is available at GEO under accession number [GSEXXXXXX](). Data is also available at Zenodo ([XX/XX]()).

The ovarian and cell derived teratoma scRNA-seq datasets included were previously published and are available at GEO ([GSEXXXXXX](), [GSEXXXXXX]()) or Zenodo ([XX/XX]()).


## Overview
This repository details the code required to replicate the transcriptomics analyses described in the manuscript - please also see the associated Methods sections from publication.


## Files
```

```


## Methods

## Quality Control and Data Preparation



## Figure 1



## Figure 2
 
 
 
## Figure 3
 
 
 
## Figure 4
 

 
## Figure 5

 
 
## Versions of packages needed:
| Package / Tool          | Version |
|-------------------------|---------|
| **R**                   | 4.3.0   |
| SoupX                   | 1.6.2   |
| Seurat                  | 5.3.0   |
| scCustomize             | 3.0.1   |
| DoubletFinder           | 2.0.4   |
| scDblFinder             | 1.16.0  |
| SingleCellExperiment    | 1.24.0  |
| ACT                     |  N/A       |
| Harmony                 | 1.2.0   |
| clustree                | 0.5.1   |
| infercnv                | 1.19.1  |
| scProportionTest        | 0.0.0.9 |
| perturbLM               | 1.0.0   |
| entropy                 | 1.3.1   |
| Semla                   | 1.1.6   |
| CARD                    | 1.1     |
| biomaRt                 | 2.58.2  |
| qvalue                  | 2.34.0  |
| msigdbr                 | 7.5.1   |
| clusterProfiler         | 4.10.1  |
| org.Hs.eg.db            | 3.18.0  |
| DESeq2                  | 1.42.1  |
| ggplot2                 | 3.5.1   |
| ggprism                 | 1.0.5   |
| circlize                | 0.4.16  |
| ComplexHeatmap          | 2.18.0  |
| ggpattern               | 1.1.1   |
| viridis                 | 0.6.5   |
| scales                  | 1.3.1   |
| ggrepel                 | 0.9.5   |
| CellRanger              | 7.0.1   |
| spaceranger             | 3.0.1   |
| LoupeBrowser            | 7.0.1   |
| cellsnp-lite            | 1.2.3   |
| Annovar                 | 2025Mar21 |



## Contact

For any questions or further information, reach out to the project lead and corresponding authors.

---
  
  *Updated on: June 18 2025*
  
